File: codes_python/run_serial.py
import numpy as np
from database import *
import pandas as pd
import scipy.cluster.hierarchy as hcluster
from wrapper import CentroidSimpleWrapper
from structures import *


# ??? toto treba niekde vhodne umiestnit je to global povodne
CENTRE_LIMIT = 0


class Serial:

    def __init__(self, args, image, log_file=""):
        self.args = args
        self.log_file = log_file
        self.image = image

    def log(self,msg):

        if self.log_file == "":
            print(msg)
        else:
            with open(self.log_file, 'a') as lf:
                print(msg, end='', file=lf)

    def clear_statistics(self):
        self.stats = Stats()

    def execute(self, index):
        self.clear_statistics()

        x_start, x_end, y_start, y_end = index


        names = ('cent.x', 'cent.y', 'snr', 'iter', 'sum', 'mean', 'var', 'std', 'skew', 'kurt', 'bckg')
        self.database  = Database(init_value=0, nrows=0, ncols=11, col_names=names )
        self.discarded = Database(init_value=0, nrows=0, ncols=11, col_names=names )

        A = self.args.width
        B = self.args.height

        if self.args.method == 'sweep':

            Xs = np.floor(np.arange(x_start + A, x_end - A, 2*A )).astype(int)
            Ys = np.floor(np.arange(y_start + B, y_end - B, 2*B )).astype(int)

            for y in Ys:
                for x in Xs:
                    self.perform_step(x, y)

        elif self.args.method == 'max':
            pixels = np.where(self.image > self.args.start_iter)

            Xs = pixels[1]
            Ys = pixels[0]

            good = (Xs > x_start) * (Xs < x_end) * (Ys > y_start) * (Ys < y_end)
            Xs = Xs[good]
            Ys = Ys[good]

            while len(Xs) > 0:
                step = self.perform_step(Xs[0], Ys[0])
                
                Xs = Xs[1:]
                Ys = Ys[1:]
                
                if step.code == 0:
                    filt_X = np.logical_and((Xs >= (step.x - A)), (Xs <= (step.x + A)))
                    filt_Y = np.logical_and((Ys >= (step.y - B)), (Ys <= (step.y + B)))
                    filt = np.logical_and(filt_X, filt_Y)
                    ok = np.logical_not(filt)

                    Xs = Xs[ok]
                    Ys = Ys[ok]

        elif self.args.method == 'cluster':
            pixels = np.where(self.image > self.args.start_iter)

            Xs = pixels[1]
            Ys = pixels[0]

            good = (Xs > x_start) * (Xs < x_end) * (Ys > y_start) * (Ys < y_end)
            Xs = Xs[good]
            Ys = Ys[good]

            pixels = np.zeros((len(Xs), 2))
            pixels[:,0] = Ys
            pixels[:,1] = Xs
            thresh = np.sqrt(A**2 + B**2)

            clusters = hcluster.fclusterdata(pixels, thresh, criterion='distance')

            for i in range(1, np.max(clusters)+1):
                XY = pixels[clusters == i]
                X = XY[:,1].astype(int)
                Y = XY[:,0].astype(int)

                Z = self.image[Y, X]

                sumG = np.sum(Z)
                sumGx = np.sum(Z*(X-0.5))
                sumGy = np.sum(Z*(Y-0.5))

                self.perform_step(sumGx/sumG, sumGy/sumG)

        return SerialResult(database=self.database, discarded=self.discarded, stats=self.stats)

    def perform_step(self, x,y):

        self.stats.started += 1

        wrapper = CentroidSimpleWrapper(image=self.image, 
                                        init_x=x, 
                                        init_y=y,
                                        A=self.args.width,
                                        B=self.args.height,
                                        noise_dim=self.args.noise_dim,
                                        alpha=self.args.angle*np.pi/180,
                                        local_noise=self.args.local_noise,
                                        delta=self.args.delta,
                                        pix_lim=self.args.start_iter,
                                        pix_prop=self.args.cent_pix_perc,
                                        max_iter=self.args.max_iter,
                                        min_iter=self.args.min_iter,
                                        snr_lim=self.args.snr_lim,
                                        fine_iter=self.args.fine_iter,
                                        is_point=self.args.width == self.args.height)
        current = wrapper.execute()

        self.update_statistics(x,y,current)
        

        if current.code == 0:
            return Step(code=0, x=current.result[0], y=current.result[1])
        else:
            return Step(code=1, x=-1, y=-1)
         
    def update_statistics(self,x,y, current):
        if current.code == 0:
            self.stats.ok += 1

            ud_code = self.database.update(current.result, CENTRE_LIMIT)

            if ud_code == 1:
                self.discarded.add(current.result)
            else:
                if self.args.verbose == 1 and self.args.parallel == 1:

                    self.log(f'{x}, {y}\n')
                    
                    for line in current.log:
                        for c in line:
                            self.log(f"{c:.6f}\t")
                        self.log('\n')
                    self.log('\n')
        
        elif current.code == 1:
            self.stats.nulldata += 1
        
        elif current.code == 2:
            self.stats.notenough += 1

        elif current.code == 3:
            self.stats.notbright += 1

        elif current.code == 4:
            self.stats.nocentre += 1

        elif current.code == 5:
            self.stats.maxiter += 1

        elif current.code == 6:
            self.stats.miniter += 1

        elif current.code == 7:
            self.stats.lowsnr += 1

        elif current.code == 8:
            self.stats.notright += 1

File: codes/script_threads/run_serial.r
# print("Tests!!!")

# print("TEST 1")
# print(ALG_PARS$CENTRE_LIMIT)
# print("TEST 2")

# single step function
performStep = function(x,y,out){
    # x,y -- where to iterate,
    # out -- if there should be output
    started <<- started + 1

    # this has to be set for the "return from function" part
    UD = list('code' = 0)

    if (x==355){
        if (y==79){
            print("debug")
        }
    }

    current =  centroidSimpleWrapper(
                IMAGE      = IMAGE,
                INIT.X     = x,
                INIT.Y     = y,
                A          = opt$width,
                B          = opt$height,
                NOISE.DIM  = opt$`noise-dim`,
                ALPHA      = opt$angle*pi/180,
                LOCALNOISE = opt$`local-noise`,
                DELTA      = opt$delta,
                PIXLIM     = opt$`start-iter`,
                PIXPROP    = opt$`cent-pix-perc`,
                MAXITER    = opt$`max-iter`,
                MINITER    = opt$`min-iter`,
                SNRLIM     = opt$`snr-lim`,
                FINEITER   = opt$`fine-iter`,
                ISPOINT    = opt$width == opt$height
                )
	
    if(current$code == 0){
        ok <<- ok + 1
        # update database if necessary
        UD = updateDatabase(DATABASE, current$result, ALG_PARS$CENTRE_LIMIT)

        DATABASE <<- UD$DATABASE
        # discarded cases
        if(UD$code == 1){
            DISCARDED <<- rbind(DISCARDED, current$result)
        }else{
            if(opt$verbose == 1 & opt$parallel == 1){
                # print iteration log for successfull runs
                cat(x); cat(', '); cat(y); cat('\n')
                for(i in 1:nrow(current$log)) {
                    for(j in 1:ncol(current$log)){
                        cat(sprintf("%.6f", current$log[i,j]))
                        cat('\t')
                    }
                    cat('\n')
                }
                cat('\n')
            }
        }
    }else if(current$code == 1){ nulldata  <<- nulldata  + 1
    }else if(current$code == 2){ notenough <<- notenough + 1
    }else if(current$code == 3){ notbright <<- notbright + 1
    }else if(current$code == 4){ nocentre  <<- nocentre  + 1
    }else if(current$code == 5){ maxiter   <<- maxiter   + 1
    }else if(current$code == 6){ miniter   <<- miniter   + 1
    }else if(current$code == 7){ lowsnr    <<- lowsnr    + 1
    }else if(current$code == 8){ notright  <<- notright  + 1
    }
    
    
    # return from function
    if(out){
        # return code and position for method max
        if(current$code == 0 & UD$code != 1){
            return(list(
                'code' = 0,
                'x'    = current$result[1],
                'y'    = current$result[2]
            ))
        }else{
            return(list(
                'code' = 1
            ))
        }
    }else{
        # no return for methods sweep and cluster
        return()
    }
}
# performStep() function updates
#   DATABASE
#   DISCARDED
#   __stat_vars__
#
# these variables come from executeSerial() serial function, so they should be visible for performStep(),
# but they will disappear when executeSerial() functions ends.
# this is important for multiple calls of executeSerial() function in parallel processing
#
# also, the performStep() function should see x.{start,end}, y.{start,end} mentioned below
#
#
# executeSerial() sees IMAGE
# executeSerial() gets what part of the image should be processed in x.{start,end}, y.{start,end}
#
# for single thread, these will be
#   x.start = 1
#   x.end   = fitsDimX
#   y.start = 1
#   y.end   = fitsDimY
#
# the variables are used in sweep/max/serial parts

executeSerial = function(index){

    # which part of the IMAGE to process
    x.start = index[1]
    x.end   = index[2]
    y.start = index[3]
    y.end   = index[4]
    
    # run statistics (__stat_vars__)
    started   = 0
    nulldata  = 0
    notenough = 0
    notbright = 0
    nocentre  = 0
    maxiter   = 0
    miniter   = 0
    lowsnr    = 0
    ok        = 0
    notright  = 0
    
    # reset database
    DATABASE  = matrix(0, nrow=0, ncol=11)
    DISCARDED = matrix(0, nrow=0, ncol=11)
    
    colnames(DATABASE)  = c('cent.x', 'cent.y', 'snr', 'iter', 'sum', 'mean', 'var', 'std', 'skew', 'kurt', 'bckg')
    colnames(DISCARDED) = c('cent.x', 'cent.y', 'snr', 'iter', 'sum', 'mean', 'var', 'std', 'skew', 'kurt', 'bckg')
    
    # box dimension
    A = opt$width
    B = opt$height
    
    # sweep through the image
    if(opt$method == 'sweep'){
        # iteration points
        Xs = floor(seq(x.start + A, x.end - A, by=2*A))
        Ys = floor(seq(y.start + B, y.end - B, by=2*B))
        for(y in Ys){
            for(x in Xs){
                environment(performStep) = environment()
                performStep(x, y, FALSE)
            }
        }
    }
    
    # find brightest points
    if(opt$method == 'max'){
        pixels = which(IMAGE > opt$`start-iter`, arr.ind = T)

        # iteration points
        Xs = pixels[,2]
        Ys = pixels[,1]

        # keep only those in range x.{start,end}, y.{start,end}
        good = (x.start < Xs) & (Xs < x.end) & (y.start < Ys) & (Ys < y.end)
        Xs = Xs[good]
        Ys = Ys[good]

        # go through all items
        while(length(Xs) > 0){
            environment(performStep) = environment()
            STEP = performStep(Xs[1], Ys[1], TRUE)
            # remove analyzed pixel
            Xs = Xs[-1]
            Ys = Ys[-1]
            # remove close pixels
            if(STEP$code == 0){
                filt_x = (Xs >= (STEP$x - A)) & (Xs <= (STEP$x + A))
                filt_y = (Ys >= (STEP$y - B)) & (Ys <= (STEP$y + B))
                Xs = Xs[!(filt_x & filt_y)]
                Ys = Ys[!(filt_x & filt_y)]
            }
        }
    }
    
    # cluster analysis
    #if(opt$method == 'clusters'){
    #    pixels = which(IMAGE > opt$`start-iter`, arr.ind = T)
    #
    #    Xs = pixels[,2]
    #    Ys = pixels[,1]
    #
    #    # keep only those in range x.{start,end}, y.{start,end}
    #    good = (x.start < Xs) & (Xs < x.end) & (y.start < Ys) & (Ys < y.end)
    #
    #    pixels = as.data.frame(pixels[good,])
    #
    #    # python clustering
    #    python.assign("pixels", pixels)
    #    python.assign("thresh", sqrt(A**2 + B**2) )
    #    python.load("script_threads/python_clustering.py")
    #    clusters = python.get("clusters")
    #
    #    pixels = cbind(pixels, clusters)
    #    pixels = pixels[order(pixels$clusters),]
    #
    #    for(i in 1:max(clusters)){
    #        # get cluster
    #        XY = pixels[pixels$clusters == i,]
    #        X  = XY[,2]
    #        Y  = XY[,1]
    #        Z  = c()
    #        for(j in 1:nrow(XY)) Z = c(Z, IMAGE[Y[j], X[j]])
    #        # centre of the cluster
    #        sumG  = sum(Z)
    #        sumGx = sum(Z*(X - 0.5))
    #        sumGy = sum(Z*(Y - 0.5))
    #        # centroiding
    #        environment(performStep) = environment()
    #        performStep(sumGx/sumG, sumGy/sumG, FALSE)
    #    }
    #
    #}
    #
    return(list(
        'DATABASE'  = DATABASE,
        'DISCARDED' = DISCARDED,
        'STATS'     = list(
            'started'   = started,
            'nulldata'  = nulldata,
            'notenough' = notenough,
            'notbright' = notbright,
            'nocentre'  = nocentre,
            'maxiter'   = maxiter,
            'miniter'   = miniter,
            'lowsnr'    = lowsnr,
            'ok'        = ok,
            'notright'  = notright
        )
    ))
}
